zero/src/prover.rs

```diff
@@ -131,29 +135,180 @@ impl BlockProverInput {
             save_inputs_on_error,
         };
 
-        // Segment the batches, prove segments and aggregate them to resulting batch
-        // proofs.
-        let batch_proof_futs: FuturesUnordered<_> = block_generation_inputs
-            .iter()
-            .enumerate()
-            .map(|(idx, txn_batch)| {
-                let segment_data_iterator =
-                    SegmentDataIterator::<Field>::new(txn_batch, Some(max_cpu_len_log));
-
-                Directive::map(IndexedStream::from(segment_data_iterator), &seg_prove_ops)
-                    .fold(&seg_agg_ops)
-                    .run(&proof_runtime.heavy_proof)
-                    .map(move |e| {
-                        e.map(|p| (idx, crate::proof_types::BatchAggregatableProof::from(p)))
-                    })
+        // Generate channels to communicate segments of each batch to a batch proving
+        // task. We generate segments and send them to the proving task, where they
+        // are proven in parallel.
+        let (segment_senders, segment_receivers): (Vec<_>, Vec<_>) = (0..batch_count)
+            .map(|_idx| {
+                let (segment_tx, segment_rx) =
+                    mpsc::channel::<Option<evm_arithmetization::AllData>>(1);
+                (segment_tx, segment_rx)
             })
-            .collect();
+            .unzip();
+
+        // The size of this channel does not matter much, as it is only used to collect
+        // batch proofs.
+        let (batch_proof_tx, mut batch_proof_rx) =
+            mpsc::channel::<(usize, crate::proof_types::BatchAggregatableProof)>(32);
+
+        // Spin up a task for each batch to generate segments for that batch
+        // and send them to the proving task.
+        let segment_generation_task = tokio::spawn(async move {
+            let mut batch_segment_futures: FuturesUnordered<_> = FuturesUnordered::new();
+
+            for (batch_idx, (txn_batch, segment_tx)) in block_generation_inputs
+                .into_iter()
+                .zip(segment_senders)
+                .enumerate()
+            {
+                batch_segment_futures.push(async move {
+                    let segment_data_iterator =
+                        SegmentDataIterator::<Field>::new(&txn_batch, Some(max_cpu_len_log));
+                    for (segment_idx, segment_data) in segment_data_iterator.enumerate() {
+                        segment_tx
+                            .send(Some(segment_data))
+                            .await
+                            .context(format!("failed to send segment data for batch {batch_idx} segment {segment_idx}"))?;
```

Comment:
idea: you could refactor all the `{batch_idx} {segment_idx}` formatting away, since you do it often.
```rust
LocatedMessage<'a> {
    msg: &'a str,
    batch: usize,
    segment: usize,
}
impl<'a> LocatedMessage<'a> {
    pub fn new(msg: &'a str, batch: usize, segment: usize) -> Self { Self {msg, batch, segment} }
}
impl fmt::Display for LocatedMessage<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Self { msg, batch, segment } = self;
        f.write_fmt(format_args!("{msg} (at batch {batch}, segment {segment})"))
    }
}
```

```suggestion
                            .context(LocatedMessage::new("failed to send segment data", batch_idx, segment_idx))?;
```

Reply:
Feels like a too big overhead and unnecessary complication.